application/cms/views.py

```diff
@@ -450,54 +441,53 @@ def send_to_review(topic_uri, subtopic_uri, measure_uri, version):
     else:
         england = wales = scotland = northern_ireland = False
 
-    form_to_validate = MeasurePageRequiredForm(
+    measure_page_form_to_validate = MeasurePageRequiredForm(
         obj=measure_page,
         meta={"csrf": False},
-        administrative_data=administrative_data,
-        survey_data=survey_data,
-        frequency_choices=FrequencyOfRelease,
-        type_of_statistic_choices=TypeOfStatistic,
         england=england,
         wales=wales,
         scotland=scotland,
         northern_ireland=northern_ireland,
         lowest_level_of_geography_choices=LowestLevelOfGeography,
     )
 
+    data_source_form_to_validate, _ = get_data_source_forms(request, measure_page=measure_page, sending_to_review=True)
+
     invalid_dimensions = []
 
     for dimension in measure_page.dimensions:
         dimension_form = DimensionRequiredForm(obj=dimension, meta={"csrf": False})
         if not dimension_form.validate():
             invalid_dimensions.append(dimension)
 
-    if not form_to_validate.validate() or invalid_dimensions:
+    measure_page_form_validated = measure_page_form_to_validate.validate()
+    data_source_form_validated = data_source_form_to_validate.validate()
+
+    if not measure_page_form_validated or invalid_dimensions or not data_source_form_validated:
         # don't need to show user page has been saved when
         # required field validation failed.
         session.pop("_flashes", None)
 
         # Recreate form with csrf token for next update
-        form = MeasurePageForm(
+        measure_page_form = MeasurePageForm(
             obj=measure_page,
-            administrative_data=administrative_data,
-            survey_data=survey_data,
-            frequency_choices=FrequencyOfRelease,
-            type_of_statistic_choices=TypeOfStatistic,
             england=england,
             wales=wales,
             scotland=scotland,
             northern_ireland=northern_ireland,
             lowest_level_of_geography_choices=LowestLevelOfGeography,
         )
 
-        for key, val in form_to_validate.errors.items():
-            form.errors[key] = val
-            field = getattr(form, key)
-            field.errors = val
-            setattr(form, key, field)
+        data_source_form, data_source_2_form = get_data_source_forms(request, measure_page=measure_page)
+
+        copy_form_errors(from_form=measure_page_form_to_validate, to_form=measure_page_form)
```

Comment:
This feels a bit weird...

Reply:
It does.

It's a symptom of the slightly weird way that we handle sending measure pages to review (and indeed also that we integrate data sources tightly into measure page flows).

It's something that should get better with time.